racket/src/expander/syntax/scope.rkt

```diff
@@ -290,6 +294,20 @@
 (define (new-scope kind)
   (scope (new-scope-id!) kind empty-binding-table))
 
+;; An "interned scope" is a scope identified by an interned symbol that is
+;; consistent across both module instantiations and bytecode unmarshalling.
+;; Creating an interned scope with the same symbol will always produce the
+;; same scope.
+(define interned-scopes-table (make-weak-hasheq))
+
+(define (make-interned-scope sym)
+  (let ([make (λ () (make-ephemeron sym (scope sym 'interned empty-binding-table)))])
+    (or (ephemeron-value
+         (hash-ref! interned-scopes-table sym make))
+        (let ([new (make)])
+          (hash-set! sym new)
+          (ephemeron-value new)))))
+
```

Comment:
Maybe this is going to be harder than I thought. The check for an existing scope and adding it to the table must be atomic, otherwise two different scopes can be returned (with the most recent one added to the table).

Reply:
Yes, that makes sense. I thought about that briefly while writing this, but didn’t fully consider the implications. Is the right thing to do here to manually lock the table with a semaphore?

Reply:
No, the expander can't have any locks that could get broken by killing a thread. You may have to resort to atomic mode, like I just did for the other table.

Reply:
Gotcha. I didn’t realize that `call-with-semaphore` would be broken by killing a thread. I guess there’s no way at all to clean up when a thread is killed short of using custodians, since `dynamic-wind` doesn’t invoke its outbound thunks on a killed thread? In any case, I can emulate what you did using atomic mode.